rust-runtime/aws-smithy-cbor/src/decode.rs

```diff
@@ -75,28 +77,85 @@
         self.decoder.skip().map_err(DeserializeError::new)
     }
 
-    // TODO An API to support both definite and indefinite strings could be:
-    //
-    //     pub fn str(&mut self, cap: u64) -> Result<&'b str, DeserializeError> {
-    //
-    pub fn str(&mut self) -> Result<&'b str, DeserializeError> {
-        self.decoder.str().map_err(DeserializeError::new)
+    // TODO-David: confirm benchmarks and keep either `str_alt` or `str`.
+    // The following seems to be a bit slower than the one we have kept.
+    pub fn str_alt(&mut self) -> Result<Cow<'b, str>, DeserializeError> {
+        // This implementation uses `next` twice to see if there is
+        // another str chunk. If there is, it returns a owned `String`.
+        let mut chunks_iter = self.decoder.str_iter().map_err(DeserializeError::new)?;
+        let head = match chunks_iter.next() {
+            Some(Ok(head)) => head,
+            None => return Ok(Cow::Borrowed("")),
+            Some(Err(e)) => return Err(DeserializeError::new(e)),
+        };
+
+        match chunks_iter.next() {
+            None => Ok(Cow::Borrowed(head)),
+            Some(Err(e)) => Err(DeserializeError::new(e)),
+            Some(Ok(next)) => {
+                let mut concatenated_string = String::from(head);
+                concatenated_string.push_str(next);
+                for chunk in chunks_iter {
+                    concatenated_string.push_str(chunk.map_err(DeserializeError::new)?);
+                }
+                Ok(Cow::Owned(concatenated_string))
+            }
+        }
+    }
+
+    pub fn str(&mut self) -> Result<Cow<'b, str>, DeserializeError> {
+        match self.decoder.str() {
+            Ok(str_value) => Ok(Cow::Borrowed(str_value)),
+            Err(e) if e.is_type_mismatch() => {
+                // Move the position back by one element to the indefinite string marker.
+                self.decoder.set_position(self.decoder.position() - 1);
```

Comment:
We're relying on the fact that `minicbor` only needs to read one byte to determine the data type here. We should document this since it's an important assumption that affects correctness.

A more robust implementation would be to cache the position before we issue `str()` and restore it afterwards.